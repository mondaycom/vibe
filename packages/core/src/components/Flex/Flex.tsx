import React, { forwardRef, useMemo, useRef } from "react";
import cx from "classnames";
import useMergeRef from "../../hooks/useMergeRef";
import Clickable from "../../components/Clickable/Clickable";
import {
  FlexAlign as FlexAlignEnum,
  FlexDirection as FlexDirectionEnum,
  FlexGap as FlexGapEnum,
  FlexJustify as FlexJustifyEnum
} from "./FlexConstants";
import { FlexDirection, FlexJustify, FlexAlign, FlexGap } from "./Flex.types";
import { ElementContent, VibeComponent, VibeComponentProps, withStaticProps } from "../../types";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import styles from "./Flex.module.scss";
import { camelCase } from "lodash-es";

export interface FlexProps extends VibeComponentProps {
  /**
   * Inline styles applied to the flex container.
   */
  style?: object;
  /**
   * The direction of the flex container.
   */
  direction?: FlexDirection;
  /**
   * The HTML element or component used as the root.
   */
  elementType?: React.ElementType;
  /**
   * If true, allows wrapping of flex items.
   */
  wrap?: boolean;
  /**
   * The content inside the flex container.
   */
  children?: ElementContent | ElementContent[];
  /**
   * Defines how flex items are aligned along the main axis.
   */
  justify?: FlexJustify;
  /**
   * Defines how flex items are aligned along the cross axis.
   */
  align?: FlexAlign;
  /**
   * The gap between flex items.
   */
  gap?: FlexGap | number;
  /**
   * The label of the flex container for accessibility.
   */
  ariaLabel?: string;
  /**
   * The tab order of the element.
   */
  tabIndex?: number;
  /**
   * Callback fired when the flex container is clicked.
   */
  onClick?: (event: React.MouseEvent) => void;
  /**
   * ID of the element describing the flex container.
   */
  ariaLabelledby?: string;
}

const Flex: VibeComponent<FlexProps> & {
  justify?: typeof FlexJustifyEnum;
  align?: typeof FlexAlignEnum;
  gaps?: typeof FlexGapEnum;
  directions?: typeof FlexDirectionEnum;
} = forwardRef(
  (
    {
      className,
      id,
      elementType = "div",
      direction = "row",
      wrap = false,
      children,
      justify = "start",
      align = "center",
      gap,
      onClick,
      style,
      ariaLabelledby,
      ariaLabel,
      tabIndex,
      "data-testid": dataTestId
    }: FlexProps,
    ref
  ) => {
    const componentRef = useRef<HTMLElement>(null);
    const mergedRef = useMergeRef(ref, componentRef);

    const gapStyle = useMemo(() => {
      if (!gap) {
        return;
      }
      if (typeof gap === "number") {
        return { gap: `${gap}px` };
      }
      return { gap: `var(--spacing-${gap})` };
    }, [gap]);

    const overrideStyle = useMemo(() => ({ ...style, ...gapStyle }), [style, gapStyle]);
    const onClickProps = useMemo(() => {
      if (onClick) return { elementType, ariaLabelledby };
      return { "aria-labelledby": ariaLabelledby };
    }, [onClick, elementType, ariaLabelledby]);
    const Element = onClick ? Clickable : elementType;

    return (
      <Element
        id={id}
        data-testid={dataTestId}
        {...onClickProps}
        ref={mergedRef}
        className={cx(
          styles.container,
          getStyle(styles, camelCase(`direction-${direction}`)),
          getStyle(styles, camelCase(`justify-${justify}`)),
          getStyle(styles, camelCase(`align-${align}`)),
          className,
          {
            [styles.wrap]: wrap
          }
        )}
        tabIndex={tabIndex}
        onClick={onClick}
        style={overrideStyle}
        aria-label={ariaLabel}
      >
        {children}
      </Element>
    );
  }
);

export default withStaticProps(Flex, {
  justify: FlexJustifyEnum,
  align: FlexAlignEnum,
  gaps: FlexGapEnum,
  directions: FlexDirectionEnum
});
