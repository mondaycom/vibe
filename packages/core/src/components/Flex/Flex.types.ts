export type FlexAlign = "start" | "center" | "end" | "stretch" | "baseline" | "initial";

export type FlexJustify = "start" | "center" | "end" | "stretch" | "space-around" | "space-between" | "initial";

export type FlexGap = "xs" | "small" | "medium" | "large";

export type FlexDirection = "row" | "column";

export type FlexGrow = "inherit" | "initial" | "revert" | "revert-layer" | "unset" | number;

export type FlexShrink = "inherit" | "initial" | "revert" | "revert-layer" | "unset" | number;

export type FlexBasis =
  | "auto"
  | "max-content"
  | "min-content"
  | "fit-content"
  | "content"
  | "inherit"
  | "initial"
  | "revert"
  | "revert-layer"
  | "unset"
  | number
  | `${number}%`;

export type FlexShorthand = {
  grow?: FlexGrow;
  shrink?: FlexShrink;
  basis?: FlexBasis;
};
