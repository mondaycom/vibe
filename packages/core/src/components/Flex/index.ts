export { default as Flex, type FlexProps } from "./Flex";
