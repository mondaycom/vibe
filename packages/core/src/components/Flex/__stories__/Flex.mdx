import { Meta } from "@storybook/blocks";
import { LIST, MENU, TABS } from "../../../storybook/components/related-components/component-description-map";
import * as FlexStories from "./Flex.stories";

<Meta of={FlexStories} />

# Flex

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Use Flex component to position group of sub-elements in one dimension, horizontal or vertical, without being dependent on a custom CSS file for positioning the sub-elements.

<Canvas of={FlexStories.Overview} />

### Import

```js
import { Flex } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Use flex component whenever you want to define a layout with one dimension.",
    "Flex layout can be either horizontal or vertical.",
    "You can defined the spacing between the layout children by using our fixed sizes - xs (4px), small (8px), medium (16px) or large( 24px)."
  ]}
/>

## Variants

### Directions

<Canvas of={FlexStories.Directions} />

### Horizontal spacing between items

<Canvas of={FlexStories.HorizontalSpacingBetweenItems} />

### Vertical spacing between items

<Canvas of={FlexStories.VerticalSpacingBetweenItems} />

### Horizontal positions

<Canvas of={FlexStories.HorizontalPositions} />

### Horizontal layout using flex

<Canvas of={FlexStories.HorizontalFlex} />

### Vertical positions

<Canvas of={FlexStories.VerticalPositions} />

### Vertical layout using flex

<Canvas of={FlexStories.VerticalFlex} />

### Support multi lines layout

You can display a layout that includes multiple lines using the flex component wrap mode.
This mode allows the layout to break into multiple lines if all the component children cannot fit into one only.

<Canvas of={FlexStories.SupportMultiLinesLayout} />

## Use cases and examples

### Flex as toolbar container

You can use flex component for create responsive toolbars

<Canvas of={FlexStories.FlexAsToolbarContainer} />

## Related components

<RelatedComponents componentsNames={[MENU, TABS, LIST]} />
