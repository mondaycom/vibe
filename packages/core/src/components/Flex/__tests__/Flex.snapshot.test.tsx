import React from "react";
import renderer from "react-test-renderer";
import Flex from "../Flex";

describe("Flex renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<Flex />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("with data-testid", () => {
    const tree = renderer.create(<Flex data-testid="test" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe("Horizontal display", () => {
    it("with children", () => {
      const tree = renderer
        .create(
          <Flex>
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with align", () => {
      const tree = renderer
        .create(
          <Flex align="end">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with justify", () => {
      const tree = renderer
        .create(
          <Flex justify="space-between">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with align", () => {
      const tree = renderer
        .create(
          <Flex align="end">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with gap", () => {
      const tree = renderer
        .create(
          <Flex gap="large">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with wrap", () => {
      const tree = renderer
        .create(
          <Flex wrap>
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with flex", () => {
      const tree = renderer
        .create(
          <Flex>
            <Flex flex={{ grow: 1, shrink: 0, basis: "auto" }}>1</Flex>
            <Flex flex={{ grow: 0, shrink: 1, basis: "auto" }}>2</Flex>
            <Flex flex={{ grow: 0, shrink: 1, basis: "auto" }}>3</Flex>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe("Vertical display", () => {
    it("with children", () => {
      const tree = renderer
        .create(
          <Flex direction="column">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with justify", () => {
      const tree = renderer
        .create(
          <Flex direction="column" justify="space-between">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with align", () => {
      const tree = renderer
        .create(
          <Flex direction="column" align="end">
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with wrap", () => {
      const tree = renderer
        .create(
          <Flex direction="column" wrap>
            <div>1</div>
            <div>2</div>
            <div>3</div>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("with flex", () => {
      const tree = renderer
        .create(
          <Flex direction="column">
            <Flex flex={{ grow: 1, shrink: 0, basis: "auto" }}>1</Flex>
            <Flex flex={{ grow: 0, shrink: 1, basis: "auto" }}>2</Flex>
            <Flex flex={{ grow: 0, shrink: 1, basis: "auto" }}>3</Flex>
          </Flex>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
