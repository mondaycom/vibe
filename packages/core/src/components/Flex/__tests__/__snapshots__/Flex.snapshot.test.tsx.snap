// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders correctly Horizontal display with align 1`] = `
<div
  className="container directionRow justifyStart alignEnd"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with align 2`] = `
<div
  className="container directionRow justifyStart alignEnd"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with children 1`] = `
<div
  className="container directionRow justifyStart alignCenter"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with flex 1`] = `
<div
  className="container directionRow justifyStart alignCenter"
  style={{}}
>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 1,
        "flexShrink": 0,
      }
    }
  >
    1
  </div>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  >
    2
  </div>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  >
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with gap 1`] = `
<div
  className="container directionRow justifyStart alignCenter"
  style={
    {
      "gap": "var(--spacing-large)",
    }
  }
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with justify 1`] = `
<div
  className="container directionRow justifySpaceBetween alignCenter"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Horizontal display with wrap 1`] = `
<div
  className="container directionRow justifyStart alignCenter wrap"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Vertical display with align 1`] = `
<div
  className="container directionColumn justifyStart alignEnd"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Vertical display with children 1`] = `
<div
  className="container directionColumn justifyStart alignCenter"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Vertical display with flex 1`] = `
<div
  className="container directionColumn justifyStart alignCenter"
  style={{}}
>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 1,
        "flexShrink": 0,
      }
    }
  >
    1
  </div>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  >
    2
  </div>
  <div
    className="container directionRow justifyStart alignCenter"
    style={
      {
        "flexBasis": "auto",
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  >
    3
  </div>
</div>
`;

exports[`Flex renders correctly Vertical display with justify 1`] = `
<div
  className="container directionColumn justifySpaceBetween alignCenter"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly Vertical display with wrap 1`] = `
<div
  className="container directionColumn justifyStart alignCenter wrap"
  style={{}}
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`Flex renders correctly with data-testid 1`] = `
<div
  className="container directionRow justifyStart alignCenter"
  data-testid="test"
  style={{}}
/>
`;

exports[`Flex renders correctly with empty props 1`] = `
<div
  className="container directionRow justifyStart alignCenter"
  style={{}}
/>
`;
